src/std/misc/repr.ss

```diff
@@ -0,0 +1,133 @@
+  :std/format :std/misc/list :std/misc/rtd :std/sort :std/srfi/1)
+
+;; Default options for printing an evaluable representation. Keep it empty for now.
+;; Note: we don't actually use options yet, but
+;; it is intended to contain pretty-printing state such as desired number of columns, etc.
+;; : Table
+(def default-representation-options (make-hash-table))
+
+;; Current options for printing an evaluable representation. Always empty for now.
+;; TODO: use immutable hash tables instead when available
+;; : (Parameter Table)
+(def current-representation-options (make-parameter default-representation-options))
+
+;; Display a list of items, with prefix and suffix, separated by given separator.
+;; Print the prefix, suffix and separator strings on the given port.
+;; Call the display-element function on each element.
+;; : <- (List Any) port: Port prefix: String separator: String suffix: String display-element: (<- Any)
+(def (display-separated
+      list
+      port: (port (current-output-port))
+      prefix: (prefix "")
+      separator: (separator " ")
+      suffix: (suffix "")
+      display-element: (display-element (lambda (x) (display x port))))
+  (display prefix port)
+  (def first? #t)
+  (for-each
+   (lambda (elem)
+     (if first?
+       (set! first? #f)
+       (display separator port))
+     (display-element elem))
+   list)
+  (display suffix port))
+
+;; Print an evaluable representation of an object on the given port with the given options.
+;; The port defaults to (current-output-port). The options are reserved for future use.
+;; : <- Any port: Port options: Table
+(def (print-representation
+      x
+      port: (port (current-output-port))
+      options: (options (current-representation-options)))
+
+  ;; Our universal utilities: print (recurse), repr, display, write
+  (def (p y) (print-representation y port: port options: options))
+  (def (r y) (repr y options: options))
+  (def (d y) (display y port))
+  (def (w y) (write y port))
+  (def (simple? x) (or (number? x) (boolean? x) (string? x) (char? x) (void? x) (eof-object? x)))
+
+  (cond
+   ;; Simplest case: just write it.
+   ((simple? x)
+    (w x))
+   ((or (symbol? x) (null? x)) ;; requires slightly more care: write it after a quote.
+    (d "'") (w x))
+   ((pair? x) ;; pair: print as [ ... ].
+    (display-separated x prefix: "[" separator: " " display-element: p port: port)
+    (let ((end (cdr (last-pair x))))
+      (cond
+       ((null? end) (void))
+       ((simple? end) (d " . ") (p end))
+       (else (d " ") (p end) (d " ..."))))
+    (d "]"))
+   ((vector? x) ;; vector: print as (vector ...).
+    (display-separated (vector->list x)
+                       prefix: "(vector " separator: " " display-element: p suffix: ")" port: port))
+   ((u8vector? x) ;; u8vector: print as (u8vector ...).
+    (display-separated (u8vector->list x)
+                       prefix: "(u8vector " separator: " " display-element: p suffix: ")" port: port))
```

Comment:
this should be quoted:
```
'#u8vector(byte ...)
```

Reply:
I don't understand. This isn't a valid form, is it?
```
> '#u8vector(1 2 3 4)
*** ERROR IN (console)@3.2 -- Invalid token
> *** ERROR IN (console)@3.11 -- Operator is not a PROCEDURE
(1 2 3 4)
```

Reply:
Sorry, typo -- this should be `'#u8(...)`